[#3] Basic configuration types and label rules

yaman opened the pull request:
Basic data types for a simple configuration.

yaman committed:
[#3] Basic configuration types and label rules

kahlil29 commented:
Seems like the build is failing. Is it compiling fine on your local machine @yaman ?

yaman commented:
@kahlil29 Yeap, it compiles fine.

kahlil29 reviewed (commented):
- Not sure whether we should be using `String` or `Text`. Of late, I have become biased towards `Text` unless some manipulations are required which would be better via `String`s. 
@chshersh or @vrom911 would be able to confirm about this.

kahlil29 commented on src/Crocodealer/Config.hs:
```diff
@@ -0,0 +1,25 @@
+data Config = Config
+  { username        :: Maybe String
+  , repository      :: Maybe String
+  , labelRules           :: [LabelRule]
```
seems like a minor formatting issue here
```suggestion
  , labelRules      :: [LabelRule]
```

yaman replied:
Seems I failed to run stylish-haskell continuously. Fixing it now..

yaman commented:
> * Not sure whether we should be using `String` or `Text`. Of late, I have become biased towards `Text` unless some manipulations are required which would be better via `String`s.
>   @chshersh or @vrom911 would be able to confirm about this.

Sure, I would be happy to change it.

yaman commented:
Build fails because;
in `Crocodealer/Config.hs`, we have our types but since we do not use them **yet**, `stack build` with flag `--ghc-options=-Werror` brakes the build.

kahlil29 commented on src/Crocodealer/Config.hs:
```diff
@@ -0,0 +1,25 @@
+  }
+
+data LabelRule
+  = Create
```
I think that this should be of the form 

```
data LabelRule = Create Label | Override <something here>
``` 

In the case of `Override`, from the example that Dmitrii has provided in the issue description, it seems like a lambda needs to be provided in order to provide the logic of what is to be done to the existing label name.
But I'm not sure how that would fit in here.

yaman replied:
I might need some guidance on this one.

chshersh replied:
@kahlil29 is right, `LabelRule` should not be just enum. It should be:

```haskell
data LabelRule
    = Create Label
    | Override (Name Label) (Name Label)
```

First element of the `Override` constructor is `from` and the second is `to`. We change only names of the labels in the `Override` and then `Create` will actually do `create or update` work.

yaman committed:
[fix] indentation

Co-Authored-By: Kahlil Abreu <kahlil29@gmail.com>

vrom911 commented on src/Crocodealer/Config.hs:
```diff
@@ -0,0 +1,25 @@
+module Crocodealer.Config
+  ( LabelRule
```
We are using this style guide over all of our projects: https://kowainik.github.io/posts/2019-02-06-style-guide
So by its rules, here we would have 7 spaces before the open bracket.

vrom911 commented on src/Crocodealer/Config.hs:
```diff
@@ -0,0 +1,25 @@
+module Crocodealer.Config
+  ( LabelRule
+  , Label
```
I guess you would need to import all of the constructors and fields for each data type here.

yaman replied:
I am not sure if I undestood this one. Can you elaborate more please Veronika?

vrom911 commented on src/Crocodealer/Config.hs:
```diff
@@ -0,0 +1,25 @@
+  ) where
+
+data Label = Label
+  { name        :: String
```
We are using 4-spaces indentation rule by our style guide.

vrom911 replied:
Also, @kahlil29 is right, it's better to use `Text` fields here and further :+1:

vrom911 replied:
Fields names should be started with the data type name prefix:
```suggestion
  { labelName        :: !Text
```

vrom911 reviewed (commented):
Nice start, @yaman ! I have some recommendations though

chshersh commented on crocodealer.cabal:
```diff
@@ -23,6 +23,7 @@ source-repository head
 library
   hs-source-dirs:      src
   exposed-modules:     Crocodealer
+                       Crocodealer.Config
```
It's not necessary required, but I prefer to indent hierarchical modules in the `.cabal` file as well for readability. See `issue-wanted` for example:

* https://github.com/kowainik/issue-wanted/blob/e612a87454b207fa47f64c630b17d014cdc4aa8a/issue-wanted.cabal#L61-L81

```suggestion
                           Crocodealer.Config
```

chshersh commented on src/Crocodealer/Config.hs:
```diff
@@ -0,0 +1,25 @@
+
+type Package = String
+
+data Config = Config
```
I would propose to separate these two concepts in two different modules:

1. `Crocodealer.Core.Label` for all label types and pure functions
2. `Crocodealer.Config` for the configuration data type and it's codec.

chshersh commented on src/Crocodealer/Config.hs:
```diff
@@ -0,0 +1,25 @@
+data Label = Label
+  { name        :: String
+  , description :: String
+  , color       :: String -- Data.Colour.sRGB24read can convert string specifications like "#00aaff" or "00aaff" 
```
We definitely should use more typesafe types here.

1. `newtype Name a = Name { unName :: Text }` and `Name Label` for the `labelName` field
2. `Text` for description is okay
3. `newtype HexColor = HexColor { unHexColor :: Text }` for `labelColor`. Constructor of `HexColor` shouldn't not exported. Instead, you need to create smart constructor:

```haskell
mkHexColor :: Text -> Maybe HexColor
```

Which verifies that the given text is in proper format: `#00ffaa` or something like that.

chshersh commented on src/Crocodealer/Config.hs:
```diff
@@ -0,0 +1,25 @@
+  = Create
+  | Override
+
+type Package = String
```
This should be: 

```haskell
newtype Repo = Repo { unRepo :: Text }
```

chshersh reviewed (changes requested):
The start is good, but some adjustments need to be made :slightly_smiling_face:

vrom911 commented:
Hey @yaman, do you need any help with the comments here?
Don't worry if you don't have time now, just let us know if we could help you :slightly_smiling_face:

yaman commented:
I was just outside of the city last week. But I'm back and ready to fix according to feedbacks :blush:

yaman committed:
[#3] applying reviewed changes to the issue;

- indenting hierarchical modules in the .cabal
- applying kowainik style-guide for module(7 spaces) definition and the rest(4 spaces)
- converting all Strings to Text
- more typesafe types
- using new types for LabelRule
- changing type Package to type Repo
- moving Label specific types under Crocodealer.Core.Label module

yaman committed:
[#3] Explicit module exports

chshersh commented on src/Crocodealer/Config.hs:
```diff
@@ -0,0 +1,17 @@
+module Crocodealer.Config
+       ( Repo(..)
+       , Config(..)
```
I usually prefer to add and extra space in between data type and its exposed constructors
```suggestion
       , Config (..)
```

chshersh commented on src/Crocodealer/Config.hs:
```diff
@@ -0,0 +1,17 @@
+       , Config(..)
+       ) where
+
+import Crocodealer.Core.Label
```
Let's add explicit import list here

chshersh commented on src/Crocodealer/Config.hs:
```diff
@@ -0,0 +1,17 @@
+       ) where
+
+import Crocodealer.Core.Label
+
```
There should be two empty lines after the import section according to our style guide
```suggestion


```

chshersh commented on src/Crocodealer/Config.hs:
```diff
@@ -0,0 +1,17 @@
+
+data Config = Config
+    { username            :: Maybe Text
+    , repository          :: Maybe Text
```
All fields in this data type should be strict. Also, they should be prefixed with the data type name. Like `configUsername` and `configRepository`

chshersh commented on src/Crocodealer/Core/Label.hs:
```diff
@@ -0,0 +1,24 @@
+data Label = Label
+    { labelName        :: Name Label
+    , labelDescription :: !Text
+    , labelColor       :: HexColor
```
All fields should be strict, not only one :slightly_smiling_face:

chshersh commented on src/Crocodealer/Core/Label.hs:
```diff
@@ -0,0 +1,24 @@
+
+data LabelRule
+    = Create Label
+    | Override (Name Label) (Name Label)
```
Would be great to add some documentation to this data type. This is the only non-trivial data type in this PR, so it's always good to add some docs to it.

vrom911 commented on src/Crocodealer/Config.hs:
```diff
@@ -0,0 +1,17 @@
+    }
+
+data Config = Config
+    { username            :: Maybe Text
```
Could we using the prefix-base naming convention for every field here? Also, we can make all the fields strickt in this data type, like here:

```suggestion
    { configUsername            :: !(Maybe Text)
```

chshersh reviewed (changes requested):
This looks nice! Code improved a lot and the structure is more clear now :slightly_smiling_face: I have some suggestions though, but we are moving forward :arrow_forward:

vrom911 commented on src/Crocodealer/Config.hs:
```diff
@@ -0,0 +1,17 @@
+module Crocodealer.Config
+       ( Repo(..)
```
There should be a space here :slightly_smiling_face: 
```suggestion
       ( Repo (..)
```

vrom911 commented on src/Crocodealer/Core/Label.hs:
```diff
@@ -0,0 +1,24 @@
+    }
+
+data Label = Label
+    { labelName        :: Name Label
```
You can make all of the fields strict

vrom911 reviewed (commented):
Nice! I think derive `Show` at least would be helpful for future work and testing.

yaman committed:
extra space in between data type and exposed structure

Co-Authored-By: Dmitrii Kovanikov <kovanikov@gmail.com>

yaman committed:
extra space in between data type and exposed structure

Co-Authored-By: Dmitrii Kovanikov <kovanikov@gmail.com>

yaman committed:
[#3] explicit imports, strict types, more descriptive data constructor names and cosmetic fixes

vrom911 reviewed (approved):
Thanks!
